tree/ntuple/v7/src/RColumnElement.hxx

```diff
@@ -740,6 +742,145 @@ public:
+   static_assert(sizeof(T) <= sizeof(double));
+   assert(1 <= nQuantBits && nQuantBits <= 8 * sizeof(Quantized_t));
+
+   const std::size_t quantMax = (1ull << nQuantBits) - 1;
+   const double scale = quantMax / (max - min);
+   const std::size_t unusedBits = sizeof(Quantized_t) * 8 - nQuantBits;
+
+   for (std::size_t i = 0; i < count; ++i) {
+      T elem = src[i];
+      assert(min <= elem && elem <= max);
+      double e = (elem - min) * scale;
+      Quantized_t q = static_cast<Quantized_t>(e + 0.5);
+      ByteSwapIfNecessary(q);
+
+      // double-check we actually used at most `nQuantBits`
+      assert(LeadingZeroes(q) >= unusedBits);
+
+      // we want to leave zeroes in the LSB, not the MSB, because we'll then drop the LSB
+      // when bit packing.
+      dst[i] = q << unusedBits;
+   }
+}
+
+/// Undoes the transformation performed by QuantizeReals() (assuming the same `count`, `min`, `max` and `nQuantBits`).
+template <typename T>
+void UnquantizeReals(T *dst, const Quantized_t *src, std::size_t count, double min, double max, std::size_t nQuantBits)
+{
+   static_assert(std::is_floating_point_v<T>);
+   static_assert(sizeof(T) <= sizeof(double));
+   assert(1 <= nQuantBits && nQuantBits <= 8 * sizeof(Quantized_t));
+
+   const std::size_t quantMax = (1ull << nQuantBits) - 1;
+   const double scale = (max - min) / quantMax;
+   const double bias = min * quantMax / (max - min);
+   const std::size_t unusedBits = sizeof(Quantized_t) * 8 - nQuantBits;
+
+   for (std::size_t i = 0; i < count; ++i) {
+      Quantized_t elem = src[i];
+      // Undo the LSB-preserving shift performed by QuantizeReals
+      assert(TrailingZeroes(elem) >= unusedBits);
+      elem >>= unusedBits;
+      ByteSwapIfNecessary(elem);
+
+      double fq = static_cast<double>(elem);
+      double e = (fq + bias) * scale;
+      dst[i] = static_cast<T>(e);
+      assert(min <= dst[i] && dst[i] <= max);
+   }
+}
+} // namespace Quantize
+
+template <typename T>
+class RColumnElementQuantized : public RColumnElementBase {
+   static_assert(std::is_floating_point_v<T>);
+
+public:
+   static constexpr bool kIsMappable = false;
+   static constexpr std::size_t kSize = sizeof(T);
+
+   RColumnElementQuantized() : RColumnElementBase(kSize, 0) {}
+
+   void SetBitsOnStorage(std::size_t bitsOnStorage) final
+   {
+      const auto [minBits, maxBits] = GetValidBitRange(EColumnType::kReal32Quant);
+      R__ASSERT(bitsOnStorage >= minBits && bitsOnStorage <= maxBits);
+      fBitsOnStorage = bitsOnStorage;
+   }
+
+   void SetValueRange(double min, double max) final
+   {
+      R__ASSERT(min >= std::numeric_limits<T>::lowest());
+      R__ASSERT(max <= std::numeric_limits<T>::max());
+      fValueRange = {min, max};
+   }
+
+   bool IsMappable() const final { return kIsMappable; }
+
+   void Pack(void *dst, const void *src, std::size_t count) const final
+   {
+      auto quantized = std::make_unique<Quantize::Quantized_t[]>(count);
```

Comment:
Perhaps add a TODO, at some point we may want to think about removing the allocation